Fix e2e decode perf regression in multi-chip Falcon7b demo and set cpu governor to performance mode in t3k demo pipeline

skhorasganiTT opened the pull request:
- A regression in e2e decode perf in the 8-chip Falcon7b demo (1400 t/s to 800 t/s) was caused by 4fc1bb4. The cause was the introduction of the ttnn concat and ttnn slice ops in the decode MLP which contain python overhead (serialized with multi-chip async).
- Fixed the regression by changing the two new ops to the tt-lib equivalents and updated the perf mode targets for the falcon7b t3000 demo
- Set the cpu governor in the t3000 demo pipeline to performance mode

skhorasganiTT committed:
#5383: [Falcon7b] Change pad/unpad in mlp to ttlib to fix multi-chip decode perf regression in demo

Signed-off-by: Salar Hosseini <skhorasgani@tenstorrent.com>

skhorasganiTT committed:
#0: Set cpu governor to performance mode for t3k demo tests

Signed-off-by: Salar Hosseini <skhorasgani@tenstorrent.com>

skhorasganiTT commented:
T3000 demo tests: https://github.com/tenstorrent/tt-metal/actions/runs/9215027315
T3000 model perf tests: https://github.com/tenstorrent/tt-metal/actions/runs/9214015838
Single-chip model perf tests: https://github.com/tenstorrent/tt-metal/actions/runs/9214012625

tt-rkim reviewed (approved):
@tapspatel 
I think the CODEOWNERS directives are wrong - we forgot to put `workflows` in the middle

We can check if changes to CODEOWNERs are working by

- going to the repo on GH
- clicking on the branch with change
- browsing to the owned filed in question
- hover over lock icon and check names on upper left corner of file. If no lock when expected, then broken

tt-rkim commented:
I second @pavlejosipovic 's questions
Was the threshold just super high before so now we're lowering it to what we want?

djordje-tt commented:
I have 2 questions:
- Is device perf impacted? 
- Can you raise an issue regarding ttnn.concat and slice overhead so we get it optimized and can get back to it instead of using tt_lib? I believe we want to use ttnn as much as possible and improve it over time since our customers will use it.

skhorasganiTT commented:
@pavlejosipovic @tt-rkim This regression was not caught in the multi-chip perf tests since for the decode tests prefill_optimized_mode is not enabled and so this path is not used. The regression would have been caught in the demo CI test (which does enable prefill_optimized_mode) but I added that in right after 4fc1bb4 (I adjusted the targets accordingly for the regression when adding it in to ensure nothing else regresses and investigated it in the mean time; now I'm readjusting the targets to the original expected perf). 

@djordje-tt Device perf is not impacted, the same underlying ttlib op was used with the ttnn versions, they just had extra python overhead that was not being parallelized for multi-chip async. Converting the python overhead to c++ is already being addressed by the metal team with the effort to push everything to ttnn. I will slowly work on converting the model to ttnn.experimental for equivalent perf and once the other ttnn issues are addressed we can convert fully to ttnn.

pavlejosipovic commented:
Should we enalbe decode test to run in it's own UT to run in the same way it would run in demo?

skhorasganiTT commented:
@pavlejosipovic We could add an additional test to the multi-chip perf pipeline for testing decode with prefill_optimized_mode enabled but now that we have the demo in CI it might be redundant.

djordje-tt commented on models/demos/falcon7b/tt/falcon_mlp.py:
```diff
@@ -288,17 +289,19 @@ def _load_mlp_padded_tensors(self):
+            and self.prefill_seq_len in [1024, 2048]
+        ):
+            for device_id in range(self.num_devices):
+                x[device_id] = tt_lib.tensor.concat(
```
Can you call ttnn.experimental instead? And the same for unpad op below?

skhorasganiTT replied:
Yes, all ttlib ops will be converted to tnn experimental in an upcoming PR, as I mentioned in the comment above.
